### PR TITLE
feat: chain-adapters `getAddress()` from unchained

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

With a new `pubKey` param - indicating public key should be used to fetch the address from unchained and bypass wallet fetching.

This allows us to effectively unrug cross-chain trades (can't have two apps open at the same time to get each address) as well as make the UX smoother, not requiring users to open the relevant app when they're getting an address i.e receive modal, trade etc. If we have xpubs / address identifiers, we can use those to either do an Identity on them (return self), or for xpubs, use unchained to get the receive address.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #

## Risk

Moderate/high, but effectively close to zero given this is isolated to Ledger. 
**Ensure that all code paths related to this are guarded against wallets other than ledger**. This param should *never* be passed to `adapter.getAddress()` for other wallets.

### Engineering

- ☝🏽 
- Put a breakpoint in chain-adapters `getAddress()` methods
- Ensure the branch never gets hit unless the current wallet is Ledger

#### Ledger flag on

- Connect your Ledger accounts
- Go to an EVM / UTXO account and click "Receive" either with the wrong app open or while on the Ledger homescreen and confirm the receive address is still working and the correct one (can be tested by importing the same seed as native in your Ledger and ensuring the receive address is the same)
  - Make sure to test with UTXO accounts that have some Tx history, i.e where the next receive/change address is *not* the zero'th 
- Cross-chain quotes are working with a Ledger connected

#### Ledger flag off

- Ensure the receive addresses aren't showing any regression. Test with native and/or KK, in particular with UTXOs and ATOM/RUNE

## Screenshots (if applicable)

image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5336** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5336" target="_blank">Graphite</a> 
    * **PR #5392** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5392" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5392?utm_source=stack-comment).

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts [hunk 1]
```diff
@@ -219,6 +220,22 @@ export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAd
 
       const bip44Params = this.getBIP44Params({ accountNumber, accountType, isChange, index })
 
+      if (pubKey) {
+        const account = await this.getAccount(pubKey)
+        const nextReceiveIndex = account.chainSpecific.nextReceiveAddressIndex
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:225]: Next receive *or* change index:
```typescript
bip44Params.isChange
          ? account.chainSpecific.nextChangeAddressIndex
          : account.chainSpecific.nextReceiveAddressIndex
      }
```

Comment (gomesalexandre) [reply packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts]: ed5b9b0d9e395286f40654a773c6d3fab75db729

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx [hunk 2]
```diff
@@ -147,6 +155,7 @@ export const useGetTradeQuotes = () => {
           affiliateBps: willDonate ? DEFAULT_SWAPPER_DONATION_BPS : '0',
           // Pass in the user's slippage preference if it's set, else let the swapper use its default
           slippageTolerancePercentage: userSlippageTolerancePercentage,
+          pubKey: fromAccountId(sellAccountId).account,
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:158]: Only if Ledger

Comment (gomesalexandre) [reply src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx]: 71e8a4f33

File: src/context/WalletProvider/Ledger/components/Chains.tsx [hunk 3]
```diff
@@ -136,23 +143,32 @@ export const LedgerChains = () => {
...
+                      <RawText fontWeight='bold'>{asset.name}</RawText>
+                    </Flex>
+                    <Button
+                      isLoading={loadingChains[asset.chainId]}
+                      onClick={() => handleConnectClick(asset.chainId)}
+                      colorScheme={walletChainIds.includes(asset.chainId) ? 'green' : 'gray'}
+                      isDisabled={walletChainIds.includes(asset.chainId)}
+                      {...(walletChainIds.includes(asset.chainId)
+                        ? { leftIcon: <CheckCircleIcon /> }
+                        : {})}
```
Comment (woodenfurniture) [inline src/context/WalletProvider/Ledger/components/Chains.tsx:165]:
```suggestion
                      leftIcon: walletChainIds.includes(asset.chainId) ? <CheckCircleIcon /> : undefined
```

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx [hunk 4]
```diff
@@ -72,10 +75,11 @@ export const useReceiveAddress = () => {
         wallet,
         accountMetadata: buyAccountMetadata,
         deviceId: await wallet.getDeviceID(),
+        ...(fetchUnchainedAddress && { pubKey: fromAccountId(buyAccountId).account }),
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:78]: nitpick but im a fan of avoid the use of object_spread for populating a single key
```suggestion
        pubKey: fetchUnchainedAddress ? fromAccountId(buyAccountId).account : undefined,
```

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx [hunk 5]
```diff
@@ -147,6 +159,7 @@ export const useGetTradeQuotes = () => {
           affiliateBps: willDonate ? DEFAULT_SWAPPER_DONATION_BPS : '0',
           // Pass in the user's slippage preference if it's set, else let the swapper use its default
           slippageTolerancePercentage: userSlippageTolerancePercentage,
+          ...maybePubKey,
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:162]: same here, the an iife+object+spread seems heavy to populate a single key 🤌
```suggestion
          pubKey: isLedger(wallet) ? fromAccountId(sellAccountId).account : undefined,
```

File: src/components/Modals/Receive/ReceiveInfo.tsx [hunk 6]
```diff
@@ -68,11 +69,17 @@ export const ReceiveInfo = ({ asset, accountId }: ReceivePropsType) => {
...
+        if (!isLedger(wallet)) return {}
+        if (!selectedAccountId) return {}
+        return { pubKey: fromAccountId(selectedAccountId as AccountId).account }
+      })()
       const selectedAccountAddress = await getReceiveAddress({
         asset,
         wallet,
         deviceId: await wallet.getDeviceID(),
         accountMetadata,
+        ...maybePubKey,
```
Comment (woodenfurniture) [inline src/components/Modals/Receive/ReceiveInfo.tsx:82]: and here too?
```suggestion
        pubKey: isLedger(wallet) && selectedAccountId 
          ? fromAccountId(selectedAccountId as AccountId).account
          : undefined,
```

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts [hunk 7]
```diff
@@ -219,6 +220,25 @@ export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAd
...
+        const account = await this.getAccount(pubKey)
+        const index = bip44Params.isChange
+          ? account.chainSpecific.nextChangeAddressIndex
+          : account.chainSpecific.nextReceiveAddressIndex
+
+        if (index === undefined)
+          throw new Error(`UtxoBaseAdapter: Could not fetch address index from unchained`)
+        if (!account.chainSpecific.addresses)
+          throw new Error(`UtxoBaseAdapter: Could not fetch addresses from unchained`)
+        const address = account.chainSpecific.addresses[index].pubkey
```
Comment (woodenfurniture) [inline packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts:233]: not sure if this is possible, but just a thought
```suggestion
        const address = account.chainSpecific.addresses[index]?.pubkey
```

**Review by woodenfurniture — COMMENTED** [review]
Conceptually lgtm, will need to pile and go through testing together.

Interestingly, while running on native wallet i got the ledger connect dialog in chrome without clicking anything. Unlikely action required as it wasnt reproducible.
image

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx (hunk 4 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/hooks/useReceiveAddress.tsx]: Might as well, as well! 9293cf6c86

File: src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx (hunk 5 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx]: Might as well ✅  12bf1a6e65

File: src/components/Modals/Receive/ReceiveInfo.tsx (hunk 6 above)
Comment (gomesalexandre) [reply src/components/Modals/Receive/ReceiveInfo.tsx]: Might as well ✅  12bf1a6e65

File: packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts (hunk 7 above)
Comment (gomesalexandre) [reply packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts]: Absolutely, makes for one less throw just for the sake of type system being happy - a336e4c5f8

**Review by woodenfurniture — APPROVED** [review]
🧠

**Pull request merged**